README.md
acmacs.r package
================

This package provides R 4.0 with ability to read, modify, save the data stored in acmacs and
lispmds files.

Supported platforms
===================

 - macOS 10.14.6

 - Linux

Dependencies
=============

Rcpp 1.0.3 or later is required.

```R
install.packages("Rcpp")
```

Installation
============

https://cran.r-project.org/web/packages/Rcpp/index.html
R 4.0 is required, other versions are not supported!

Package is supposed to be installed from a binary archive for the
specific platform. Building package from sources is complicated and it
cannot be done within R.

Installation on macOS 10.14
---------------------

```R
install.packages(c("Rcpp", "devtools", "testthat", "roxygen2"))
install.packages(Find(function(asset) { grepl("macOS", asset$name, fixed=TRUE) }, devtools:::github_GET("repos/acorg/acmacs.r/releases/latest")$assets)$browser_download_url)
```


Installation on Linux
-----

```R
install.packages(c("Rcpp", "devtools", "testthat", "roxygen2"))
install.packages(Find(function(asset) { grepl("linux", asset$name, fixed=TRUE) }, devtools:::github_GET("repos/acorg/acmacs.r/releases/latest")$assets)$browser_download_url)

```

Loading
---------------------

```R
library(acmacs.r)
```

Usage example
=============

Complete example with the most of available functions can be accessed within R using ?acmacs.Example

```R
library(acmacs.r)
chart <- new(acmacs.Chart, "/tmp/file.ace")
cat(chart$info)

ags <- chart$antigens
format_antigen <- function(a) {
  paste(c(a$name, a$passage, a$reassortant, a$reference, a$annotations, paste("[", a$date, "]", sep="", collapse=""), a$lab_ids), collapse=" ", sep="")
}
cat("antigens:", length(ags), "\n ", paste(sapply(ags, format_antigen), collapse="\n  "), "\n")
```

Docs
=======

[Rcpp](https://cran.r-project.org/web/packages/Rcpp/index.html)


NEWS
acmacs.r NEWS -- history of user-visible changes.

* Changes in 5.1 (2020-12-03)
** Old variants of chart$relax calls use dimension annealing
** New variants of chart$relax calls (with options) have dimension annealing turned off by default

* Changes in 5.0 (2020-10-12)
** Dimension annealing is off by default
** chart$relax(<number-of-dimeansions>, <option> ...)

* Changes in 4.6 (2020-06-04)
** Using R 4.0 on macOS

* Changes in 4.5 (2020-05-17)
** bug fix in inserting/removing antigens (in acmacs-chart-2)

* Changes in 4.4 (2020-04-12)
** new chart$relax_incremental interface
merge$relax_incremental(1000)
merge$relax_incremental(1000, FALSE) - the same as merge2$relax_incremental(1000, "rough")
merge$relax_incremental(1000, "rough", "unmovable-primary-points") - freeze primary map points

* Changes in 4.3 (2020-03-20)
** set step for grid test
grid_test <- new(acmacs.GridTest, chart, 0.25) - with step 0.25
grid_test <- new(acmacs.GridTest, chart, 1, 0.25) - for the second best projection with step 0.25
** dimension annealing
chart <- new(acmacs.Chart, filename)
chart$remove_all_projections()
projection <- chart$new_projection("none", 5) # initially make 5D projection
projection$relax()
projection$dimension_annealing(2)  # convert relaxed projection to 2D
projection$relax()  # relax in 2D
** set disconnected, unmovable points
projection$set_disconnected(<vector of point indexes>)
projection$set_unmovable(<vector of point indexes>)
projection$set_unmovable_in_the_last_dimension(<vector of point indexes>)

* Changes in 4.2 (2020-03-05)
** Minor fix in hope to debug strange behaviour on some machines

* Changes in 4.1 (2020-03-04)
** Shared lib installation issue under Linux fixed
** Color name handling improvements. If a color is semi-transparent, its value returned in the css rgba() form
** Support for setting and getting fill and outline color opacity
   chart$plot_spec$set_outline(1, "red")
   chart$plot_spec$styles[[1]]$outline
     [1] "red"
   chart$plot_spec$set_outline_opacity(1, 0.1)
   chart$plot_spec$styles[[1]]$outline_opacity
     [1] 0.1019608
   chart$plot_spec$styles[[1]]$outline
     [1] "rgba(255,0,0,0.102)"

* Changes in 4.0 (2019-12-20)
** chart$new_projection_with_layout(minimum_column_basis, matrix(...)), columns of the matrix corresponds to the dimensions of the layout
** optimization with the seeded layout randomization: chart$relax("1280", 2, FALSE, seed)
** acmacs.merge.1() .. acmacs.merge.5(), see https://github.com/acorg/acmacs-chart-2/blob/master/doc/merge-types.org
** chart$relax_incremental()
** map resolution test: specify number of optimization: acmacs.map_resolution_test(chart=chart1, number_of_dimensions=2:3, proportions_to_dont_care=c(0.1, 0.5), number_of_optimizations=1000)
** map resolution test: specify number of replicates: acmacs.map_resolution_test(chart=chart1, number_of_random_replicates_for_each_proportion=25)
** map resolution test: specify directory to save intermediate charts: acmacs.map_resolution_test(chart=chart1, save_charts_to="/path/to/directory")

* Changes in 3.6 (2019-07-10)
** bug fix in procrustes for layouts with common disconnected points

* Changes in 3.5 (2019-07-01)
** bug fix in merging titers for multi-layered source charts

* Changes in 3.4 (2019-06-09)
** bug fix (resetting transformation upon relaxing a projection)
** bug fix, avoid running grid test for points having no titers at all

* Changes in 3.3 (2019-06-05)
** chart$extension_field() and chart$set_extension_field()

* Changes in 3.2 (2019-05-24)
** chart$save() returns string with chart saved in ace format (uncompressed)
** trying to relax projection when chart has no titers leads to an exception (randomizer_plain_from_sample_optimization_internal: diameter is 0)

* Changes in 3.1 (2019-05-13)
** chart$plot_spec$styles_as_data_frame()
** chart$plot_spec$set_styles(<data-frame returned by styles_as_data_frame())
** chart$plot_spec$set_outline_width(), set_shown(), set_rotation(), set_aspect(), set_shape(), set_label_shown(), set_label_offset_x(), set_label_offset_y(), set_label_size(), set_label_color(), set_label_rotation(), set_label_slant(), set_label_weight(), set_label_font_family(), set_label_text()

* Changes in 3.0 (2019-05-10)
** Workaround for libomp mutliple initialiation issue and corresponding crash
** Checking titer value when it is set (negative and non-integer titers forbidden)
** chart$name returns just name, can be empty if it was not set
** chart$column_bases() method takes a 1-based projection number
** use projection$set_transformation(matrix(...)), projection$transformation <- matrix(...) does not work (bug in Rcpp?)
** error message about locationdb absence suppressed
** set the column basis for each serum individually
** applying procrustes (transformation and translation) to layout (R matrix)
** getting titers of layers
** matching antigens and sera in two charts
** stress object access for making custom optimisers
** map resolution test

* Changes in 2.99.1 (2019-04-30)
** Avoiding libomp double initialisation error by using libomp coming with R under macOS
** Using R 3.6 and Rcpp 1.0.1

* Changes in 2.11 (2018-11-27)
** getting transformation of projection (2D and 3D)
** setting transformation of projection (2D and 3D)
** procrustes for just antigens or just sera
** reorient projection of a chart to a master projection (of the same or different chart)

* Changes in 2.10 (2018-11-13)
** merging tables
** incremental merge
** frozen merge
** overlay merge
** titer merge diagnostics

* Changes in 2.9 (2018-11-07)
** randomizing layout of a projection (projection$randomize_layout())
** running optimization step by step (projection$relax_one_iteration())

* Changes in 2.8 (2018-10-19)
** bug fix in reading layered charts made by acmacs-web


* Changes in 2.7 (2018-10-17)
** R 3.5.1
** single thread variant of the grid test
** remove all projections in the chart (chart$remove_all_projections())

tests/test-omp.R
library(acmacs.r)

rs <- function() {
    rs_ <- rowSums(data.frame(a=c(3,3,3),b=c(4,4,4)))
    cat("\nrowSums executed: ")
    cat(rs_)
    cat("\n\n")
}

rs()
chart <- new(acmacs.Chart, "testthat/cdc-h1pdm-20090612.ace")
cat("chart loaded\n")
cat(chart$info)
cat("\n\nabout to call chart$relax_many()\n\n")
chart$relax_many("none", 2, 10, TRUE)
cat(chart$info)
rs()


tests/testthat.R
library(testthat)
library(acmacs.r)
test_check("acmacs.r")


src/grid-test.cpp
#include "acmacs-base/fmt.hh"
#include "acmacs-chart-2/grid-test.hh"

#include "grid-test.hh"

// ----------------------------------------------------------------------

GridTest::GridTest(Chart& chart, size_t projection_no, double grid_step) :
    grid_test_(new acmacs::chart::GridTest(*chart.obj_, projection_no, grid_step))
{
}

// ----------------------------------------------------------------------

Rcpp::DataFrame GridTest::test_parallel()
{
    results_ = std::make_shared<acmacs::chart::GridTest::Results>(grid_test_->test_all());
    return results();
}

// ----------------------------------------------------------------------

Rcpp::DataFrame GridTest::test_single_thread()
{
    results_ = std::make_shared<acmacs::chart::GridTest::Results>(grid_test_->test_all(1));
    return results();
}

// ----------------------------------------------------------------------

Projection GridTest::make_new_projection_and_relax()
{
    if (results_) {
        if (results_->count_trapped_hemisphering() == 0)
            throw std::runtime_error("test found neither trapped nor hemisphering points, nothing to do");
        auto projection = grid_test_->make_new_projection_and_relax(*results_, false);
        results_.reset();
        grid_test_->reset(projection);
        return projection;
    }
    else
        throw std::runtime_error("run test first");
}

// ----------------------------------------------------------------------

Rcpp::DataFrame GridTest::results()
{
    if (results_) {
        const auto size = results_->count_trapped_hemisphering();
        const auto num_dims = results_->num_dimensions();
        Rcpp::NumericVector v_point_no(size);
        Rcpp::StringVector v_diagnosis(size);
        Rcpp::NumericVector v_pos_x(size);
        Rcpp::NumericVector v_pos_y(size);
        Rcpp::NumericVector v_pos_z(size);
        Rcpp::NumericVector v_distance(size);
        // Rcpp::NumericVector v_contribution_diff(size);

        size_t index = 0;
        for (const auto& result : *results_) {
            if (result) {
                v_point_no[index] = result.point_no + 1;
                v_diagnosis[index] = result.diagnosis_str();
                v_pos_x[index] = result.pos.x();
                v_pos_y[index] = result.pos.y();
                if (num_dims > acmacs::number_of_dimensions_t{2})
                    v_pos_z[index] = result.pos.z();
                v_distance[index] = result.distance;
                // v_contribution_diff[index] = result.contribution_diff;
                ++index;
            }
        }
        switch (*num_dims) {
            case 0:
                return Rcpp::DataFrame::create(Rcpp::Named("point_no") = v_point_no);
            case 1:
                return Rcpp::DataFrame::create(Rcpp::Named("point_no") = v_point_no, Rcpp::Named("diagnosis") = v_diagnosis, Rcpp::Named("pos_x") = v_pos_x,
                                               Rcpp::Named("distance") = v_distance); //, Rcpp::Named("contribution_diff") = v_contribution_diff);
            case 2:
                return Rcpp::DataFrame::create(Rcpp::Named("point_no") = v_point_no, Rcpp::Named("diagnosis") = v_diagnosis, Rcpp::Named("pos_x") = v_pos_x, Rcpp::Named("pos_y") = v_pos_y,
                                               Rcpp::Named("distance") = v_distance); //, Rcpp::Named("contribution_diff") = v_contribution_diff);
            default:
                return Rcpp::DataFrame::create(Rcpp::Named("point_no") = v_point_no, Rcpp::Named("diagnosis") = v_diagnosis, Rcpp::Named("pos_x") = v_pos_x, Rcpp::Named("pos_y") = v_pos_y,
                                               Rcpp::Named("pos_z") = v_pos_z,
                                               Rcpp::Named("distance") = v_distance); //, Rcpp::Named("contribution_diff") = v_contribution_diff);
        }
    }
    else {
        Rcpp::NumericVector nv;
        return Rcpp::DataFrame::create(Rcpp::Named("point_no") = nv);
    }
}

// ----------------------------------------------------------------------
/// Local Variables:
/// eval: (if (fboundp 'eu-rename-buffer) (eu-rename-buffer))
/// End:


src/titers.cpp
#include "titers.hh"

// ----------------------------------------------------------------------

Rcpp::StringMatrix Titers::all() const
{
    Rcpp::StringMatrix result(obj_->number_of_antigens(), obj_->number_of_sera());
    for (size_t ag_no = 0; ag_no < obj_->number_of_antigens(); ++ag_no)
        for (size_t sr_no = 0; sr_no < obj_->number_of_sera(); ++sr_no)
            result(ag_no, sr_no) = static_cast<std::string>(obj_->titer(ag_no, sr_no));
    return result;
}

// ----------------------------------------------------------------------

Rcpp::List Titers::all_layers() const
{
    if (obj_->number_of_layers() < 2)
        throw std::invalid_argument("chart has no layers");
    const auto number_of_antigens = obj_->number_of_antigens(), number_of_sera = obj_->number_of_sera();
    const auto& data = obj_->layers();
    Rcpp::List result(data.size());
    for (size_t layer_no = 0; layer_no < data.size(); ++layer_no) {
        Rcpp::StringMatrix layer_data(number_of_antigens, number_of_sera);
        std::fill(layer_data.begin(), layer_data.end(), "*");
        for (size_t ag_no = 0; ag_no < number_of_antigens; ++ag_no) {
            for (auto [sr_no, titer] : data[layer_no][ag_no])
                layer_data(ag_no, sr_no) = *titer;
        }
        result[layer_no] = std::move(layer_data);
    }
    return result;

} // Titers::all_layers

// ----------------------------------------------------------------------

Rcpp::StringVector Titers::titers_for_all_layers(size_t ag_no, size_t sr_no) const
{
    if (obj_->number_of_layers() < 2)
        throw std::invalid_argument("chart has no layers");
    if (ag_no < 1 || ag_no > obj_->number_of_antigens())
        throw std::invalid_argument("invalid antigen_no");
    if (sr_no < 1 || sr_no > obj_->number_of_sera())
        throw std::invalid_argument("invalid serum_no");
    const auto titers = obj_->titers_for_layers(ag_no - 1, sr_no - 1, acmacs::chart::Titers::include_dotcare::yes);
    Rcpp::StringVector result;
    for (auto titer : titers)
        result.push_back(*titer);
    return result;

} // Titers::titers_for_all_layers

// ----------------------------------------------------------------------
/// Local Variables:
/// eval: (if (fboundp 'eu-rename-buffer) (eu-rename-buffer))
/// End:


src/acmacs.cpp
#include <Rcpp.h>

#include "locationdb/locdb.hh"

// #include <limits>
// #include "chart.hh"

// ----------------------------------------------------------------------

// inline auto style_shown(acmacs::PointStyle* style) { return *style->shown; }
// inline auto style_size(acmacs::PointStyle* style) { return style->size->value(); }
// inline std::string style_fill(acmacs::PointStyle* style) { return style->fill->to_string(); }
// inline std::string style_outline(acmacs::PointStyle* style) { return style->outline->to_string(); }
// inline auto style_outline_width(acmacs::PointStyle* style) { return style->outline_width->value(); }
// inline auto style_rotation(acmacs::PointStyle* style) { return style->rotation->value(); }
// inline auto style_aspect(acmacs::PointStyle* style) { return style->aspect->value(); }
// inline std::string style_shape(acmacs::PointStyle* style) { return *style->shape; }

// inline Rcpp::StringVector passage_as_character(acmacs::chart::Passage* aPassage) { return {*aPassage}; }
// inline Rcpp::List Chart::get_antigens() const { return getList<Antigen>(obj_->antigens_modify()); }
// inline Rcpp::List Chart::get_sera() const { return getList<Serum>(obj_->sera_modify()); }

// ----------------------------------------------------------------------

static inline void init_cout_cerr()
{
    get_locdb(locdb_suppress_error::yes); // to avoid error message if locdb is requested later and locationdb.json.xz is not available
    std::cout.rdbuf(Rcpp::Rcout.rdbuf());
    std::cerr.rdbuf(Rcpp::Rcerr.rdbuf());
}

// ----------------------------------------------------------------------

RCPP_MODULE(acmacs)
{
    using namespace Rcpp;

    function("acmacs.init_cout_cerr", &init_cout_cerr, "automatically called on module loading");

    // class_<Chart>("acmacs.Chart")
    //         .constructor<std::string>("read chart data from a file", &Chart::validate_constructor<std::string>)
    //         .constructor<Rcpp::RawVector>("read chart from raw inline data", &Chart::validate_constructor<Rcpp::RawVector>)
    //         .constructor<int, int>("create new chart from scratch", &Chart::validate_constructor<int>)
    //         .property<size_t>("number_of_antigens", &Chart::number_of_antigens)
    //         .property<size_t>("number_of_sera", &Chart::number_of_sera)
    //         .property<size_t>("number_of_points", &Chart::number_of_points)
    //         .property<size_t>("number_of_projections", &Chart::number_of_projections)
    //         .property<std::string>("lineage", &Chart::lineage)
    //         .property<std::string>("info", &Chart::info, "multi-line string brifly describing data stored in the chart")
    //         .property<std::string>("name", &Chart::name, &Chart::set_name)
    //         .property<Rcpp::List>("antigens", &Chart::get_antigens)
    //         .property<Rcpp::List>("sera", &Chart::get_sera)
    //         .method("insert_antigen", &Chart::insert_antigen)
    //         .method("insert_serum", &Chart::insert_serum)
    //         .method("remove_antigens", &Chart::remove_antigens)
    //         .method("remove_sera", &Chart::remove_sera)
    //         .property<Rcpp::List>("projections", &Chart::getListViaAt<Projection, &acmacs::chart::ChartModify::projections_modify>)
    //         .property<PlotSpec>("plot_spec", &Chart::plot_spec)
    //         .property<Titers>("titers", &Chart::titers)
    //         .method("column_bases", &Chart::column_bases_2, "either forced or calculated column bases, i.e. the ones used for stress calculation and optimization")
    //         .method("column_bases", &Chart::column_bases_1, "either forced or calculated column bases, i.e. the ones used for stress calculation and optimization")
    //         // .method("column_bases", &Chart::column_bases_1s, "either forced or calculated column bases with the provided minimum_column_basis")
    //         .method("column_bases", &Chart::column_bases_0, "either forced or calculated column bases, i.e. the ones used for stress calculation and optimization")
    //         .method("save", &Chart::save)
    //         .method("relax", &Chart::relax2)
    //         .method("relax", &Chart::relax3)
    //         .method("relax_many", &Chart::relax_many, "generate maps multiple times from random starts\n\targuments:\n\tminimum column basis, e.g. \"none\", \"1280\"\n\tnumber of dimensions, e.g. 2\n\tnumber of optimizations, e.g. 10\n\tuse rough optimization (30% faster): TRUE or FALSE\n")
    //         .method("sort_projections", &Chart::sort_projections)
    //         ;
    // function("as.character.Rcpp_acmacs.Chart", &Chart::as_character, "as.character.Rcpp_acmacs.Chart(chart) - shows brief info about passed chart");

    // class_<Antigen>("acmacs.Antigen")
    //         .property("name", &Antigen::get_name, &Antigen::set_name, "antigen name")
    //         .method("set_name", &Antigen::set_name)
    //         .property("full_name", &Antigen::get_full_name)
    //         .property("abbreviated_name", &Antigen::get_abbreviated_name)
    //         .property("date", &Antigen::get_date, &Antigen::set_date)
    //         .method("set_date", &Antigen::set_date)
    //         .property<acmacs::chart::Passage>("passage", &Antigen::get_passage)
    //         .method("set_passage", &Antigen::set_passage)
    //         .property("lineage", &Antigen::get_lineage, &Antigen::set_lineage)
    //         .method("set_lineage", &Antigen::set_lineage)
    //         .property("reassortant", &Antigen::get_reassortant, &Antigen::set_reassortant)
    //         .method("set_reassortant", &Antigen::set_reassortant)
    //         .property("reference", &Antigen::get_reference, &Antigen::set_reference)
    //         .method("set_reference", &Antigen::set_reference)
    //         .property<Rcpp::StringVector>("lab_ids", &Antigen::get_lab_ids)
    //         .property<Rcpp::StringVector>("annotations", &Antigen::get_annotations)
    //         .method("add_annotation", &Antigen::add_annotation)
    //         .method("remove_annotation", &Antigen::remove_annotation)
    //         ;
    // function("as.character.Rcpp_acmacs.Antigen", &Antigen::as_character);

    // class_<Serum>("acmacs.Serum")
    //         .property("name", &Serum::get_name, &Serum::set_name)
    //         .method("set_name", &Serum::set_name)
    //         .property("full_name", &Serum::get_full_name)
    //         .property("abbreviated_name", &Serum::get_abbreviated_name)
    //         .property("passage", &Serum::get_passage)
    //         .method("set_passage", &Serum::set_passage)
    //         .property("lineage", &Serum::get_lineage, &Serum::set_lineage)
    //         .method("set_lineage", &Serum::set_lineage)
    //         .property("reassortant", &Serum::get_reassortant, &Serum::set_reassortant)
    //         .method("set_reassortant", &Serum::set_reassortant)
    //         .property("serum_id", &Serum::get_serum_id, &Serum::set_serum_id)
    //         .method("set_serum_id", &Serum::set_serum_id)
    //         .property("serum_species", &Serum::get_serum_species, &Serum::set_serum_species)
    //         .method("set_serum_species", &Serum::set_serum_species)
    //         .property<Rcpp::StringVector>("annotations", &Serum::get_annotations)
    //         .method("add_annotation", &Serum::add_annotation)
    //         .method("remove_annotation", &Serum::remove_annotation)
    //         ;
    // function("as.character.Rcpp_acmacs.Serum", &Serum::as_character);

    // class_<acmacs::chart::Passage>("acmacs.Passage")
    //         .method("is_egg", &acmacs::chart::Passage::is_egg)
    //         .method("is_cell", &acmacs::chart::Passage::is_cell)
    //         .method("without_date", &acmacs::chart::Passage::without_date)
    //         .method("type", &acmacs::chart::Passage::passage_type)
    //         ;
    // function("as.character.Rcpp_acmacs.Passage", &passage_as_character);

    // class_<Projection>("acmacs.Projection")
    //         .property<std::string>("info", &Projection::make_info)
    //         .property("stress", &Projection::stress)
    //         .method("recalculate_stress", &Projection::recalculate_stress)
    //         .property<std::string>("comment", &Projection::comment)
    //         .property<std::string>("minimum_column_basis", &Projection::minimum_column_basis)
    //         .property("forced_column_bases", &Projection::forced_column_bases)
    //         .property("transformation", &Projection::transformation)
    //         .property("layout", &Projection::layout)
    //         .property("transformed_layout", &Projection::transformed_layout)
    //         .method("rotate_degrees", &Projection::rotate_degrees)
    //         .method("rotate_radians", &Projection::rotate_radians)
    //         .method("flip_east_west", &Projection::flip_east_west)
    //         .method("flip_north_south", &Projection::flip_north_south)
    //         .method("move_point", &Projection::move_point)
    //         .method("relax", &Projection::relax)
    //         .method("relax", &Projection::relax_default)
    //         ;

    // class_<PlotSpec>("acmacs.PlotSpec")
    //         .property("styles", &PlotSpec::styles)
    //         .property("drawing_order", &PlotSpec::drawing_order)
    //         .method("raise", &PlotSpec::drawing_order_raise)
    //         .method("lower", &PlotSpec::drawing_order_lower)
    //         .method("raise_sera", &PlotSpec::drawing_order_raise_sera)
    //         .method("lower_sera", &PlotSpec::drawing_order_lower_sera)
    //         .method("set_size", &PlotSpec::set_style_size)
    //         .method("set_fill", &PlotSpec::set_style_fill)
    //         .method("set_outline", &PlotSpec::set_style_outline)
    //         ;

    // class_<acmacs::PointStyle>("acmacs.PointStyle")
    //         .property("shown", &style_shown, nullptr)
    //         .property("size", &style_size, nullptr)
    //         .property("fill", &style_fill, nullptr)
    //         .property("outline", &style_outline, nullptr)
    //         .property("outline_width", &style_outline_width, nullptr)
    //         .property("rotation", &style_rotation, nullptr)
    //         .property("aspect", &style_aspect, nullptr)
    //         .property("shape", &style_shape, nullptr)
    //           // LabelStyle label;
    //           // field<std::string> label_text;
    //         ;

    // class_<Titers>("acmacs.Titers")
    //         .method("all", &Titers::all)
    //         .method("titer", &Titers::titer)
    //         .method("set_titer", &Titers::set_titer)
    //         .method("set_dontcare_for_antigen", &Titers::set_dontcare_for_antigen)
    //         .method("set_dontcare_for_serum", &Titers::set_dontcare_for_serum)
    //         .method("multiply_by_for_antigen", &Titers::multiply_by_for_antigen)
    //         .method("multiply_by_for_serum", &Titers::multiply_by_for_serum)
    //         ;

    // function("acmacs.procrustes", &procrustes);

    // class_<ProcrustesData>("acmacs.ProcrustesData")
    //         .property("rms", &ProcrustesData::rms, nullptr)
    //         .property("transformation", &ProcrustesData::transformation, nullptr)
    //         ;

}

// ----------------------------------------------------------------------
/// Local Variables:
/// eval: (if (fboundp 'eu-rename-buffer) (eu-rename-buffer))
/// End:
